cleanlab/datalab/internal/issue_manager/underperf_group.py

```diff
@@ -0,0 +1,222 @@
+        self.info = self.collect_info(
+            knn_graph=knn_graph, n_clusters=n_clusters, cluster_labels=cluster_labels
+        )
+
+    def _process_knn_graph_from_inputs(self, kwargs: Dict[str, Any]) -> Union[csr_matrix, None]:
+        """Determine if a knn_graph is provided in the kwargs or if one is already stored in the associated Datalab instance."""
+        knn_graph_kwargs: Optional[csr_matrix] = kwargs.get("knn_graph", None)
+        knn_graph_stats = self.datalab.get_info("statistics").get("weighted_knn_graph", None)
+
+        knn_graph: Optional[csr_matrix] = None
+        if knn_graph_kwargs is not None:
+            knn_graph = knn_graph_kwargs
+        elif knn_graph_stats is not None:
+            knn_graph = knn_graph_stats
+
+        if isinstance(knn_graph, csr_matrix) and kwargs.get("k", 0) > (
+            knn_graph.nnz // knn_graph.shape[0]
+        ):
+            # If the provided knn graph is insufficient, then we need to recompute the knn graph
+            # with the provided features
+            knn_graph = None
+        return knn_graph
+
+    def collect_info(
+        self, knn_graph: csr_matrix, n_clusters: int, cluster_labels: np.ndarray
+    ) -> dict:
+
+        params_dict = {
+            "metric": self.metric,
+            "k": self.k,
+            "threshold": self.threshold,
+        }
+
+        N = knn_graph.shape[0]
+        dists = knn_graph.data.reshape(N, -1)[:, 0]
+        nn_ids = knn_graph.indices.reshape(N, -1)[:, 0]
+
+        knn_info_dict = {
+            "nearest_neighbor": nn_ids.tolist(),
+            "distance_to_nearest_neighbor": dists.tolist(),
+        }
+
+        statistics_dict = self._build_statistics_dictionary(knn_graph=knn_graph)
+
+        cluster_stat_dict = self._get_cluster_statistics(n_clusters=n_clusters)
+        info_dict = {**params_dict, **knn_info_dict, **statistics_dict, **cluster_stat_dict}
+
+        return info_dict
+
+    def _build_statistics_dictionary(self, knn_graph: csr_matrix) -> Dict[str, Dict[str, Any]]:
+        statistics_dict: Dict[str, Dict[str, Any]] = {"statistics": {}}
+
+        # Add the knn graph as a statistic if necessary
+        graph_key = "weighted_knn_graph"
+        old_knn_graph = self.datalab.get_info("statistics").get(graph_key, None)
+        old_graph_exists = old_knn_graph is not None
+        prefer_new_graph = (
+            not old_graph_exists
+            or knn_graph.nnz > old_knn_graph.nnz
+            or self.metric != self.datalab.get_info("statistics").get("knn_metric", None)
+        )
+        if prefer_new_graph:
+            statistics_dict["statistics"][graph_key] = knn_graph
+            if self.metric is not None:
+                statistics_dict["statistics"]["knn_metric"] = self.metric
+
+        return statistics_dict
+
+    def _get_cluster_statistics(self, n_clusters: int) -> Dict[str, Dict[str, Any]]:
+        cluster_stats = {"HDBSCAN": {"n_clusters": n_clusters}}
```

Comment:
I'd like to suggest a slight modification in how we structure the clustering information. We can make the dictionary more intuitive and maintainable, especially if we decide to expand on it or integrate other clustering methods in the future.

```suggestion
        cluster_stats = {
            "clustering": {
                "algorithm": "HDBSCAN",  # or any other method we might use in the future
                "params": {
                    # Specific parameters related to the chosen clustering algorithm
                    "min_cluster_size": 5,
                    "min_samples": None,
                    "metric": "precomputed",
                    # ... any other parameters that may be useful ...
                },
                "stats": {
                    # Derived statistics or post-clustering results
                    "n_clusters": n_clusters,
                    "cluster_labels": labels_,
                    "silhouette_score": sklearn.metrics.silhouette_score(...)
                    # ... any other statistics/results ...
                }
            }
        }
```